File: LiveDataSampleJava/app/src/main/java/com/websarva/wings/android/livedatasamplejava/ui/MainViewModel.java
package com.websarva.wings.android.livedatasamplejava.ui;

import android.app.Application;

import com.websarva.wings.android.livedatasamplejava.data.CocktailmemoRepository;
import com.websarva.wings.android.livedatasamplejava.data.local.Cocktailmemo;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

public class MainViewModel extends AndroidViewModel {
	private CocktailmemoRepository _cocktailmemoRepository;
	private MutableLiveData<Cocktailmemo> _cocktailmemoLiveData = new MutableLiveData<>();

	public MainViewModel(Application application) {
		super(application);
		_cocktailmemoRepository = new CocktailmemoRepository(application);
		Cocktailmemo cocktailmemo = new Cocktailmemo();
		cocktailmemo.id = -1;
		cocktailmemo.name = "未選択";
		cocktailmemo.note = "";
		_cocktailmemoLiveData.setValue(cocktailmemo);
	}

	public void prepareCocktailmemo(int cocktailId, String cocktailName) {
		Cocktailmemo cocktailmemo = _cocktailmemoRepository.getCocktailmemo(cocktailId);
		if(cocktailmemo == null) {
			cocktailmemo = new Cocktailmemo();
			cocktailmemo.id = cocktailId;
			cocktailmemo.name = cocktailName;
			cocktailmemo.note = "";
		}
		_cocktailmemoLiveData.setValue(cocktailmemo);
	}

	public void saveCocktailmemo(String note) {
		Cocktailmemo cocktailmemo = _cocktailmemoLiveData.getValue();
		cocktailmemo.note = note;
		boolean result = _cocktailmemoRepository.deleteInsertCocktailmemo(cocktailmemo);
		if(result) {
			cocktailmemo.id = -1;
			cocktailmemo.name = "未選択";
			cocktailmemo.note = "";
			_cocktailmemoLiveData.setValue(cocktailmemo);
		}
	}

	public LiveData<Cocktailmemo> getCocktailmemoLiveData() {
		return _cocktailmemoLiveData;
	}
}

File: LiveDataSampleKotlin/app/src/main/java/com/websarva/wings/android/livedatasamplekotlin/ui/MainViewModel.kt
package com.websarva.wings.android.livedatasamplekotlin.ui

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.websarva.wings.android.livedatasamplekotlin.data.CocktailmemoRepository
import com.websarva.wings.android.repositorysamplekotlin.data.local.Cocktailmemo
import kotlinx.coroutines.launch

class MainViewModel(application: Application) : AndroidViewModel(application) {
	private val _cocktailmemoRepository: CocktailmemoRepository
	val cocktailmemoLiveData = MutableLiveData<Cocktailmemo>()

	init {
		_cocktailmemoRepository = CocktailmemoRepository(application)
		val cocktailmemo = Cocktailmemo(-1, "未選択", "")
		cocktailmemoLiveData.value = cocktailmemo
	}

	fun prepareCocktailmemo(cocktailId: Int, cocktailName: String) {
		viewModelScope.launch {
			var cocktailmemo = _cocktailmemoRepository.getCocktailmemo(cocktailId)
			if(cocktailmemo == null) {
				cocktailmemo = Cocktailmemo(cocktailId, cocktailName, "")
			}
			cocktailmemoLiveData.value = cocktailmemo
		}
	}

	fun saveCocktailmemo(note: String) {
		var cocktailmemo = Cocktailmemo(cocktailmemoLiveData.value!!.id, cocktailmemoLiveData.value!!.name, note)
		viewModelScope.launch {
			val result = _cocktailmemoRepository.deleteInsertCocktailmemo(cocktailmemo)
			if(result) {
				cocktailmemo = Cocktailmemo(-1, "未選択", "")
				cocktailmemoLiveData.value = cocktailmemo
			}
		}
	}
}
